### ISSUE TITLE
Input format error with msc-generator files

### ISSUE THREAD

**Issue by Turbo-Diesel** [doxygen/doxygen#7994]
Hello, 

after the update from doxygen version 1.8.15 to 1.8.20, i got errors with including of msc-generator (*.signalling files) over the command \mscfile.

**The produced error log are below:**
Input format error for 'D:/Software/............./Init.signalling'
error: Problems generating msc output (error=INPUT FORMAT ERROR). Look for typos in you msc file  D:/Software/....../Init.signalling
Error detected at line 1: syntax error, unexpected '=', expecting '{'.

**Content of the signalling file is: (not very special...)**
msc=rose;
hscale="auto";
Applik, T1;
|||;
activate Applik;
Applik -> T1 :AAAA() [url="\ref AAAA"];
deactivate Applik;
activate T1;           
deactivate T1;
T1>>Applik :ReturnStatus; 
Applik -> T1 :BBBBB() [url="\ref BBBBB "];
|||;

thank you in advance

**Comment by doxygen** [comment]
Please attach a self-contained example (sources + config file) in a zip or tar so we can actually reproduce the problem.

**Comment by Turbo-Diesel** [comment]
Hello, below the attached example.

[DoxygenTest.zip](https://github.com/doxygen/doxygen/files/5160287/DoxygenTest.zip)

**Comment by albert-github** [comment]
With mscgen 0.20 I could not even reproduce your example in doxygen 1.8.15.

@Turbo-Diesel Which version of mscgen are you using?

**Comment by Turbo-Diesel** [comment]
I use the msc-generator 6.3.8
https://sourceforge.net/projects/msc-generator/files/msc-generator/v6.x/
With the MSC Generator program i save the chart into a *.signalling file. This file i add it with \mscfile command from doxygen in markdown file.

In Doxygen version 1.8.15 its works without errors. After the update to version 1.8.20, i get this errors.

![Doxygen error](https://user-images.githubusercontent.com/70583928/91975386-a1a5fb00-ed1f-11ea-9f68-7cd38a15ff87.png)
![SignallingFileFromMSC_Generator_6_3_8](https://user-images.githubusercontent.com/70583928/91975391-a2d72800-ed1f-11ea-8a40-9a2e4e90847a.png)

**Comment by albert-github** [comment]
As far as I can see in the original definition of mscgen the term "activate" is not present and this

**Comment by Turbo-Diesel** [comment]
Yes thats true. But why it works with doxygen 1.8.15 without problems with the signalling file? In the doxygen release notes from version 1.8.16 i see a lot of changes with "msc". 
What i know is, since 1.8.16 is mscgen integrated in doxygen.

**Comment by Turbo-Diesel** [comment]
Which other kind of msc files (data types) are available in doxygen with the \mscfile command?

**Comment by albert-github** [comment]
My previous (now removed comment was finished by accident.


As far as I can see in the original definition of mscgen the term "activate" is not present (http://www.mcternan.me.uk/mscgen/) and as this version was not maintained anymore so the original mscgen code was moved into doxygen as some fixes were needed. This move also means that external versions like version 6.3.8 from Sourceforge won't work anymore. Probably the Sourceforge version also added possibilities like 'activate' to the language.

An alternative is to use plantuml (see command `\startuml`). The syntax is a little bit different but your image would probably be like:
```
@startuml
participant Applik
participant T1
activate Applik
Applik -> T1 :AAAA()
deactivate Applik
activate T1
T1 --> Applik :ReturnStatus
deactivate T1
Applik -> T1 :BBBBB()
@enduml
```

![plantuml_example](https://user-images.githubusercontent.com/5223533/91980162-21839380-ed27-11ea-9c5b-bd588cdcbb23.jpg)

**Comment by Turbo-Diesel** [comment]
Is it possible for the next update of doxygen to difference the internal mscgen tool between a external msc-generator tool?
In way of: If msc install path is edited then use the external tool. If msc install path isn't edited use internal mscgen tool from doxygen self?

Otherwise the best solution for me is to downgrade Doxygen back to 1.8.15, before if i edit more than 30+ signalling files...

**Comment by LinuxIsBetter** [comment]
I switched from mscgen to msc-generator a few years ago.  Better features, -more forgiving syntax and a GUI front-end so non-markup-oriented team members could contribute/edit sequence diagrams too.

Being able to specify an external tool for msc file compiling made this possible.  The external tools also work well for quickly rendering while editing to catch errors. I haven't had very good luck getting doxygen to provide errors for msc sequences, but perhaps that's because I haven't been using the built-in engine.

Compatibility with msc-generator would be nice -- either integrating with Doxygen or returning the Doxyfile setting for choosing an external/optional compiler.

I use plantuml as well, but it doesn't support hyperlinks to code (the last time I checked).

**Comment by tlwott** [comment]
This issue is blocking also our team to move to a newer Doxygen version, as inline msc snippet would stop building.

Would it be possible to add a different `\mscgenerator` (and `\mscgeneratorfile` maybe?) tag which reinstates the old behavior of calling out to an external binary?

I gave a naive try at reverting patch 6fb61b2c6561149c1a318df26fd1acac619b3519 on top of v1.9.2, to try and live with a custom build of Doxygen, but the the surrounding code has moved too much.

**Comment by albert-github** [comment]
@tlwott Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by tlwott** [comment]
@albert-github attached the example case [msc_testcase.tar.gz](https://github.com/doxygen/doxygen/files/7502838/msc_testcase.tar.gz). `tar xf msc_testcase.tar.gz && cd msc_testcase && ./runme.sh` to reproduce.

You can peek at the already existing doxygen.log from a previous build and see how with 1.8.13 all goes smooth (because it calls out to the `msc` binary to render msc content), while with 1.9.2 you get
```
Error detected at line 5: syntax error.
> USR: User code {
Input format error for 'msc_testcase/doxygen_1.9.2/latex/inline_mscgraph_1.msc'
msc_testcase/Readme.txt:60: error: Problems generating msc output (error=INPUT FORMAT ERROR). Look for typos in you msc file msc_testcase/doxygen_1.9.2/latex/inline_mscgraph_1.msc
```
because Doxygen it tries to render msc content itself.

You might need to adapt the reproduction script to your environment. From `runme.sh`:
```
# README.
# Assumptions, which you might have to alter for your working environment:
# - which version of Doxygen to run can be magically selected with the DOXYGEN_VERSION variable
# - here Doxygen 1.8.13 and 1.9.2 are used to show the difference; anything pre/post-1.8.16 should show the same difference in behaviour
# - there exists a msc binary in the PATH that actually points to msc-generator from https://sourceforge.net/projects/msc-generator/
```

**Comment by albert-github** [comment]
It looks to me that the problem is that the `msc-generator` has some more features than the original `mscgen` and the doxygen implementation.
For doxygen it was decided not use to the `mscgen` executable anymore as it is not maintained anymore. In the doxygen `libmscgen` directory  we see in the README.txt:
> Issue 6880 (https://github.com/doxygen/doxygen/issues/6880) required fixes to the mscgen tool.
Since this tool is no longer maintained, it was decided to build mscgen as part of doxygen.
This directory contains the mscgen code.

When we run the original `mscgen` executable with the 1.8.13 version (Mscgen version 0.20, Copyright (C) 2010 Michael C McTernan, Michael.McTernan.2001@cs.bris.ac.uk) we don't see a message either but no image is created (the messages are suppressed). Running the `mscgen` directly on the msc file we get:
```
Error detected at line 5: syntax error, unexpected characters, expecting ',' or ';'.
> USR: User code {`
```

When running with the 1.9.2 version we get:
```
.../Readme.txt:60: error: Problems generating msc output (error=INPUT FORMAT ERROR). Look for typos in you msc file .../html/inline_mscgraph_1.msc

Error detected at line 5: syntax error, unexpected characters, expecting ',' or ';'.
> USR: User code {
```
so in fact the same message though now shown properly.

There are a number of possibilities:
- add an option to the doxygen configuration with which it is again possible to define ones own generator for handling the `\msc` / `\mscfile` code (though I could not find a Windows way to build the executable, the build process looks geared to the *nix way of building).
- see what the extensions are and implement these into the current `libmscgen`, I think this is quite a big task and beyond the scope and possibilities of doxygen.
- try to use `msc-generator` as a library, though I don't know where this will lead to, there might be some problems with used (graphical) libraries (and I didn't look whether it is allowed in terms of the license)
- An alternative might be to use PlantUML: https://plantuml.com/

@doxygen what is your opinion?

**Comment by tlwott** [comment]
Waiting for a more official view on the possible solutions, my vote would be to add a new `\mscgenerator` / `\mscgeneratorfile` tags (or equivalent) that do exactly what `\msc` / `\mscfile` used to do, that is call out to the external tool binary for rendering of the msc contents.
This way retrocompatibility is maintaned, expectations of users currently using `\msc` are not broken and who wants to use the richer `msc-generator` is not at a dead end.

**Comment by teknos293** [comment]
Hi, I am the author of msc-generator. I have forked from mscgen 14 years ago exactly to add more features - but to keep backwards compatibility as mscgen was integrated with a lot of tools already. Msc-generator uses cairo for the drawing and while it is possible to compile it into doxygen, perhaps it is simpler to use as an external program for those who need its extra features.  #9057 is a patch directly doing that. However, I am willing to collaborate on whatever solution is best for doxygen.

**Comment by albert-github** [comment]
The pull request https://github.com/doxygen/doxygen/pull/9059 for the issue https://github.com/doxygen/doxygen/issues/9057 has been integrated into the master version of doxygen.
@tlwott, @LinuxIsBetter,  Does this solve your issues ?

**Comment by LinuxIsBetter** [comment]
I cloned the repo and tried a test build on windows following instructions at https://www.doxygen.nl/manual/install.html
using command
 _cmake -G "Visual Studio 15 2017" c:\tools\testBuild\doxygen\src_
and got errors.
Unable to build. Unable to answer. Sorry.

**Comment by albert-github** [comment]
@LinuxIsBetter thanks for the attempt.

I'm not sure your configuration command is correct, the mentioned install page says:
```
mkdir build
cd build
cmake -G "Visual Studio 15 2017" ..
```
and it looks like you try either to build in place or you mentioned the incorrect directory (should be in your case be: `c:\tools\testBuild\doxygen`
- What errors did you get?
- Wnot use an executable from the Artifacts on GitHub Actions like from https://github.com/doxygen/doxygen/actions/runs/4397228881?

**Comment by LinuxIsBetter** [comment]
@albert-github  Thanks very much for the quick/helpful response!

**Re: Windows build attempt,
the instructions I actually used were from the BUILD.txt in the repo, which says:**
```
For Windows one can generate a Visual Studio project using
  cmake -G "Visual Studio 12 2013" path\to\root\of\doxygen\source\tree
```
I guessed (incorrectly) that path/to/.../source/tree must be the src/ directory.
I can provide errors later if needed, although it's possible the version of Visual Studio I'm using is not supported.

**Re: Wnot use an executable from the Artifacts on GitHub Actions**
GitHub actions are new to me.  Thanks for the tip!  Windows Chrome refused to download "Windows Latest MSVC Release build artifacts" claiming that it may be dangerous, but I was able to download the Debug counterpart and install it.

I created a test.dox file with a "@mainpage" line followed by an "@mscfile <path>" line for each of the mscfiles in the folder with my Doxyfile, and used 
`MSCGEN_TOOL            = "c:/Program Files (x86)/Msc-generator/mscgen.exe"`

Results are:
- only the first successful "@mscgen" line is handled, and subsequent "@mscgen" lines are ignored

More debugging in progress...

**Comment by LinuxIsBetter** [comment]
Today I started with small examples and methodically worked my way up.
**Things look great.**
I'm sure the one-and-done error I reported is a mistake on my part.

I also noted that warnings generated by mscgenerator are reported in the Doxygen output -- very helpful.
Thanks!

**Comment by albert-github** [comment]
Thanks for testing.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.7.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
